### DISCUSSION TITLE
Authorization and API requests

### DISCUSSION THREAD

**Discussion by jarvis394** [jarvis394/geekr#69] [General]
На данный момент _(09.12)_ готова авторизация пользователя через почту и пароль. Чтобы делать запросы на API Хабра, нужно использовать прокси, т.к. их API (https://habr/api/v1) не позволяет по какой-то причине делать запросы с мобильным **User-Agent**. Короче говоря, с телефона нельзя делать запросы, поэтому я сделал [маленький прокси](https://github.com/jarvis394/habra-backend) и уже с него будут идти все реквесты с токенами.

**Проблема 1:** хотелось бы иметь доступ к API с телефона (из браузера), но для этого нужно либо разбираться с самим Хабром, либо каким-то образом менять заголовок **User-Agent** (что вроде как сделать нельзя). Иметь прокси - это довольно плохая затея, так как а) передаётся приватный токен и б) будет большая нагрузка на маленький сервер по части количества запросов.

**Проблема 2:** документация к API Хабра довольно скудна. Если кто найдет что-нибудь, то кидайте сюда. Пока есть:
- https://github.com/Perkovec/HabrApi/tree/master
- https://github.com/Makentoshe/Habrachan
- https://gist.github.com/jarvis394/71e2533a721ee0da7d03dd66a2bf535c

Сейчас я делаю новый дизайн и перевожу приложение на работу с новым API.

**Comment by centralhardware** [comment]
отличные новости

**Comment by centralhardware** [comment]
@jarvis394  круто, что авторизация заработала, однако какое то аномальное количество запросов к 
![image](https://user-images.githubusercontent.com/26829359/102619443-0bcb9a00-416f-11eb-843c-28e03c697a57.png)

**Reply by jarvis394** [reply]
Эти запросы делает библиотека [react-detect-offline](https://www.npmjs.com/package/react-detect-offline). Когда в браузере нет поддержки `navigator.online`, она по интервалу делает запросы на сайт.

> [1] Polling is only used as a fallback for browsers that don't support the "online" event. Currently these are Chrome on Windows, Firefox on Windows, and Chrome on Linux.

**Comment by Perkovec** [comment]
@jarvis394 Приятно видеть, что кому-то понадобились мои заметки по апи хабра (https://github.com/Perkovec/HabrApi/tree/master). Дайте знать, если все еще нуждаетесь в документированных апи методах, буду пополнять эту импровизированную документацию.

**Reply by jarvis394** [reply]
> буду пополнять эту импровизированную документацию

Было бы очень здорово! Сейчас есть проблема такого характера: 

1. Есть авторизованный клиент и неавторизованный. Оба клиента делают запрос для получения каких-либо данных, но в первом случае придут данные с дополнительными полями (+ с совершенно другой структурой), а в другом приходят другие данные.
2. Нужно написать библиотеку, которая:
    - на вход получает метод, опционально токен
    - делает запрос на тот или иной endpoint, основываясь на наличии токена
    - форматирует данные под одну структуру и отдаёт клиенту
3. Клиент получает от библиотеки данные в одном формате, который можно типизировать и радуется.

Проблема в написании этой библиотеки: как привести данные с двумя разными структурами в один вид? Делать таблицу, где поле одной структуры соответствует полю в отформатированной? Есть ли примеры таких библиотек?
Можно ли вообще избежать этого? Как я смотрел, `https://habr.com/api/v1` отвечает 400 на любой запрос без токена.

В общем, пока всё неясно с авторизацией и хабром, я лучше буду дизайнить новую версию приложения и пилить фичи, чем костылить всё это дело. Может, раз вы знаете что-то про их апи, расскажете про более хорошие методы получения информации в одном виде, в одной структуре.
